plugins/catalog/src/state/EntityListProvider.tsx

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useReducer } from 'react';
+import { entityListContext } from './context';
+import { entityListInitialState, entityListReducer } from './reducer';
+
+export const EntityListProvider = ({
+  children,
+}: {
+  children?: React.ReactNode;
+}) => {
+  const [state, dispatch] = useReducer(
+    entityListReducer,
+    entityListInitialState,
+  );
```

Comment:
So, this isn't actually using redux at all. It's still using a common react `useReducer` + `useContext` combination. There is no global redux state tree and it's isolated to this context. I chose to use the [redux-toolkit](https://redux-toolkit.js.org/) as a way to reduce common boiler plate for creating typed actions/reducers. So even though the library has redux in the name, it's just applying patterns and there is not dependency on redux itself. I can absolutely see the confusion and I could also remove the usage of `redux-toolkit`, but it would introduce a lot more boilerplate and ceremony around create actions & implementing the reducers.

As to the concern of introducing `useReducer` / redux pattern, I agree this is a little against the grain. It's was something that @freben suggested to as an attemp help simplify the complex filtering logic. I did run into a bit of friction when trying to leverage existing hooks. Mostly this was when there was no pure function equivalent of the logic for hooks. You can see an example of where I needed to extract [isStarredEntity](https://github.com/backstage/backstage/pull/2791/files#diff-4b38392a5959f94c0a7fa4743cfbfd26c18d20908e7947dbb91c3985abdce8d1R27) out of `useStarredEntities`. However, for the most part I was able to just wrap all the other hooks in the `useEntityListState` hook and it worked out decently well.

So ... there is still a big question of whether this pattern (or something similar) makes sense to use here and possibly elsewhere when hooks are the norm so far. I will say that I am not a hook expert, but I've seen some pretty complex hooks in the throughout codebase that are very hard to reason about imo. I think hooks can work well with discipline, but they can also get out of hand (probably like most patterns). I do think what is here, is actually quite simply to grok if you understand reducers and a little bit of redux. I would also argue, testing reducers is so much easier. However, this still begs a much bigger question around what makes sense for the catalog filters and the codebase in general. One of the reasons, I'm holding off on deeper changes and tests until this can be decided/agreed on.

Reply:
Thanks for the clarification, Andrew! My mistake conflating `redux` and `redux-toolkit`, I saw the redux dependency and leapt a bit to conclusions. I'd advocate for not using `redux-toolkit` but it's possible that would end up way uglier. 😅  I could fiddle around and PR the PR so we have a straw man to attack? Search filtering is definitely a bit complex by nature.

Have you thought about pagination? I looked into it briefly in the main code but thought it might be better to add on top of this PR.

Reply:
So I am noticing `redux` is pulled as a dependency with `redux-toolkit` even though it's not required or used directly. Will need to look into that. I do removing it would be a lot more verbose and less succinct. There are a few alternative libraries I've used like [typescript](https://github.com/aikoven/typescript-fsa) & [typescript-fas-reducers](https://github.com/dphilipson/typescript-fsa-reducers) which would probably be more lightweight and for sure has no direct dep on redux. Happy to also compare what it would look like with no sugar'd helpers.

> Have you thought about pagination? I looked into it briefly in the main code but thought it might be better to add on top of this PR.

Thought about it a little, but trying to keep the scope down as it's not something implemented in the current version (at least from the backend). Completely agree this is needed. Looks like some folks have 20k entitites, so this absolutely will be needed. The thinking here is there will be both client side and server side filters that are combined. 

My initial idea is that `EntityFilter` are generalized, but are either client or server based. When one is applied there is a side affect to compute the full filter set which is composed of a server predicate & client predicate. If the server version changes (checked through some equality check) it will do a fetch to the server and then apply the client filter set to the backend result. At least this is the high level idea, but a fair ways to get there. Hoping that can be done incrementally.

Reply:
After a little research into modern typing of actions/reducers, I ended up finding a decent approach without any libraries. I've removed the usage of `redux-toolkit` as a result.